### ISSUE TITLE
Open Graph protocol and link sharing

### ISSUE THREAD

**Issue by hayesall** [sphinx-gallery/sphinx-gallery#707]
It would be cool if there was a configuration option to support Open Graph protocol or Twitter cards when sharing examples:

- [Open Graph protocol](https://ogp.me/): "*enables any web page to become a rich object in a social graph.*"
- [Twitter Cards](https://developer.twitter.com/en/docs/tweets/optimize-with-cards/overview/abouts-cards): "*... you can attach rich photos, videos and media experiences to Tweets.*"

I think this could be accomplished by adding relevant `meta` tags to the head when generating pages.

### Current version

Using `scikit-learn` as an example, sharing the [`linear_model`](https://scikit-learn.org/stable/modules/linear_model.html) and [`plot_comparison`](https://scikit-learn.org/stable/auto_examples/classification/plot_classifier_comparison.html) pages on Slack results in the following:

![image](https://user-images.githubusercontent.com/11916674/84408763-891fbc00-abda-11ea-982f-a07fc71ab25d.png)

But the [generated image](https://scikit-learn.org/stable/_images/sphx_glr_plot_classifier_comparison_001.png), description, and a few other things could also be part of this.

### Possible version

Now using [this web page](https://hayesall.com/blog/latex-automata/) as an example, sharing a link results in:

![image](https://user-images.githubusercontent.com/11916674/84409445-6d68e580-abdb-11ea-9cb4-52b2b00586c5.png)

... because of the following `meta` tags:

```html
<!-- Open Graph -->
<meta property="og:locale" content="en">
<meta property="og:type" content="article">
<meta property="og:title" content="LaTeX Finite Automata and State Diagrams with Tikz">
<meta property="og:description" content="This short guide collects some examples of drawing finite state machines using the Tikz LaTeX library.">
<meta property="og:url" content="https://hayesall.com/blog/latex-automata/">
```

### Other notes

- `OPG` support might be incorporated elsewhere (e.g. [`sphinx-contrib/ogp`](https://github.com/sphinx-contrib/ogp)).
- Here's a liquid template for generating some of these: https://github.com/hayesall/hayesall.github.io/blob/fee47f82f5cb489f14cbe8018cf812733670138c/_includes/open-graph.html

**Comment by larsoner** [comment]
@hayesall sounds reasonable to me, seems like something we could add in a `.. raw:: html` in our generated RST files without too much work. Would you be up for trying to implement it?

How do you envision the API? It seems like:
```
sphinx_gallery_conf = {
    ...
    social_meta_tags=True,
}
```
could be a minimal verison for OGP if we assume types `en/article` and get `title`/`description`/`url` like we do for the builds/example pages anyway. As long as the `og:url` can be a relative path it seems like it should work without too much effort. And if people want more fields filled in (like twitter), we could allow `dict` as:
```
sphinx_gallery_conf = {
    ...
    social_meta_tags={'site.owner.twitter': ...},
}
```
or whatever other fields are needed to support twitter, etc.?

**Comment by hayesall** [comment]
> Would you be up for trying to implement it?

If you can assign me to this issue I can schedule some time! It probably won't be for a few weeks though (I have to get through some higher priorities first, but the use-case seems pretty obvious and it seems like an option that other people might want as well).

Your API suggestion is really close to what I had in mind.

There could also be some options for individual pages. Default values probably need some thought too:

- Disable social tags on individual pages: `# sphinx_gallery_social_meta_tags = False`
- Choose a specific image? (maybe default to `sphinx_gallery_thumbnail_number`)
- Disable `twitter:image` on pages where no images are generated?

**Comment by choldgraf** [comment]
I think the one thing we should double-check is that `meta` tags for opengraph work outside of the `<head>` of a page. If they do, then I think this is a good idea. If they don't, then it might be trickier to get this working across different themes

**Comment by larsoner** [comment]
> There could also be some options for individual pages. 

Unless you already know that you will want these, I'd rather add them later if people request them and assume YAGNI for now.

> Choose a specific image? (maybe default to sphinx_gallery_thumbnail_number)

Let's just use `sphinx_gallery_thumbnail_number` -- it should be representative for the gallery page, so should be good enough for Twitter sharing, too (until proven otherwise).

> Disable twitter:image on pages where no images are generated?

Agreed, should only add images where possible (the default/no-image-generated that we use in the gallery in this case is necessary just for uniformity of the gallery page itself).

> If they don't, then it might be trickier to get this working across different themes

If it must be in `<head>`, perhaps we can add a little JS code snippet (using `sphinx_app.add_js_file(...)`) to every page that looks for these `<meta>` tags and removes them from the `<body>` to `<head>`. Not sure if Twitter's page scraping would look for such dynamically created content or not ... ?

If not, we'll have to dig deeper into how to add content to the `<head>`. There might be some way of making Sphinx inject some HTML in the head from the RST, but a very quick search did not turn anything up...

**Comment by rowanc1** [comment]
I believe that the meta tags [have to be in the head](https://stackoverflow.com/questions/1447842/what-happens-if-the-meta-tags-are-present-in-the-document-body#:~:text=META%20tags%20are%20only%20allowed,essentially%20creating%20an%20invalid%20markup.) (at least last time I did this) as well as in the static server rendered html. There is a nice way to check that they are working across sites here:
* https://metatags.io/

**Comment by nabobalis** [comment]
Sorry for bring up such an old issue.

I ran into this problem and wanted to embed this information for our documentation and someone linked me: https://github.com/wpilibsuite/sphinxext-opengraph

This extension works on the gallery pages but the issue is that due to the relative paths that the gallery produces, opengraph doesn't resolve the image path in previews.

```
<meta property="og:image" content="../../../_images/sphx_glr_downloading_lascoC2_001.png" />
```

Is it possible to work around this via configuration?

**Comment by lucyleeow** [comment]
Thanks for your comment @nabobalis, potentially if sphinxext-opengraph works well, it would be a better solution than us implementing it.

Sphinx Gallery generated rst gives an image directive with a relative path to the image, e.g., from our an example in our docs:

```
.. image:: /auto_examples/images/sphx_glr_plot_0_sin_001.png
    :alt: plot 0 sin
    :class: sphx-glr-single-img
```
when Sphinx generates the html, it looks like this:
```
<img alt="plot 0 sin" class="sphx-glr-single-img" src="../_images/sphx_glr_plot_0_sin_001.png" />
```

so it looks like sphinxext-opengraph is pulling the image info from the html page generated, does this seem right for your example? Is there a way to set the config for sphinxext-opengraph such that you can specify a pattern and do some string formatting to get the image url? For example the image above has the following url:
```
https://sphinx-gallery.github.io/stable/_images/sphx_glr_plot_0_sin_001.png
```

**Comment by nabobalis** [comment]
Thanks for the follow up @lucyleeow.

Yeah that looks like that is the case, I think I will raise an issue upstream and see what they say about it.

**Comment by GaelVaroquaux** [comment]
@lucyleeow : when upstream makes progress, we should try using "sphinxext-opengraph" in the sphinx-gallery docs :)

**Comment by lucyleeow** [comment]
Yes will do! I could even mention it in our docs. Would be great if this works as avoids us from adding it and then having to maintain it.

**Comment by GaelVaroquaux** [comment]
I've just added it in dirty-cat, and it seems to work quite well:
The link https://dirty-cat.github.io/stable/auto_examples/06_dimension_reduction_and_performance.html#sphx-glr-auto-examples-06-dimension-reduction-and-performance-py
Renders as such in mattermost:
![image](https://user-images.githubusercontent.com/208217/123958657-99553580-d9ad-11eb-8219-a3be35d4f662.png)

**Comment by lucyleeow** [comment]
Looking at the inspect element of that page, it seems that image is a relative path:
```
<img alt="06 dimension reduction and performance" class="sphx-glr-single-img" src="../_images/sphx_glr_06_dimension_reduction_and_performance_001.png">
```

but it seems to resolve..!

**Comment by nabobalis** [comment]
(I think this is a difference in how discourse and mattermost handles the relative paths)
